src/Avalonia.Vulkan/Controls/VulkanControlBase.cs

```diff
@@ -0,0 +1,242 @@
+                }
+
+            _attachment.Image.TransitionLayout(ImageLayout.ColorAttachmentOptimal, AccessFlags.AccessColorAttachmentReadBit);
+        }
+
+        private void DoCleanup()
+        {
+            if (_platformInterface != null)
+            {
+                try
+                {
+                    if (_initialized)
+                    {
+                        _initialized = false;
+                        OnVulkanDeinit(_platformInterface, new VulkanImageInfo(_attachment.Image as VulkanImage));
+                    }
+                }
+                finally
+                {
+                    _platformInterface.Device.QueueWaitIdle();
+                    _attachment?.Dispose();
+                    _oldAttachment?.Dispose();
+                    _oldAttachment = null;
+                    _attachment = null;
+                }
+            }
+        }
+
+        protected override void OnDetachedFromVisualTree(VisualTreeAttachmentEventArgs e)
+        {
+            DoCleanup();
+            base.OnDetachedFromVisualTree(e);
+        }
+
+        private bool EnsureInitializedCore()
+        {
+            _platformInterface = AvaloniaLocator.Current.GetService<VulkanPlatformInterface>();
+            if (_platformInterface == null || _platformInterface.Device == null)
+            {
+                // Device is not ready. Try to initialize on the next render
+                Dispatcher.UIThread.InvokeAsync(InvalidateVisual, DispatcherPriority.Render);
+                return false;
+            }
+
+            lock (_platformInterface.Device.Lock)
+            {
+                try
+                {
+                    EnsureTextureAttachment();
+
+                    return true;
+                }
+                catch (Exception e)
+                {
+                    Logger.TryGet(LogEventLevel.Error, "Vulkan")?.Log("VulkanControlBase",
+                        "Unable to initialize Vulkan Image: {exception}", e);
+                    return false;
+                }
+            }
+        }
+
+        private bool EnsureInitialized()
+        {
+            if (_initialized)
+                return true;
+            _initialized = EnsureInitializedCore();
+
+            if (!_initialized)
+                return false;
+
+            OnVulkanInit(_platformInterface, new VulkanImageInfo(_attachment.Image));
+
+            return true;
+        }
+        
+        private PixelSize GetPixelSize()
+        {
+            var scaling = VisualRoot.RenderScaling;
+            return new PixelSize(Math.Max(1, (int)(Bounds.Width * scaling)),
+                Math.Max(1, (int)(Bounds.Height * scaling)));
+        }
+
+
+        protected virtual void OnVulkanInit(VulkanPlatformInterface platformInterface, VulkanImageInfo info)
+        {
+            
+        }
+
+        protected virtual void OnVulkanDeinit(VulkanPlatformInterface platformInterface, VulkanImageInfo info)
+        {
+            
```

Comment:
Cosmetic: use ```{ }``` for empty functions or leave a comment ```// nothing to do``` as many code analysis scanners would mention that as an issue